File: web/lite/src/base/transform-2d.ts
import { OPAQUE } from '../constant';
import { Point, ZERO_POINT } from './point';
import { Property } from './property';

import type { TimeRange } from './time-range';

export class Transform2D {
  public static createDefaultTransform2D() {
    const transform = new Transform2D();

    return transform;
  }

  public anchorPoint: Property<Point>; // spatial
  public position: Property<Point>; // spatial
  public xPosition: Property<number>;
  public yPosition: Property<number>;
  public scale: Property<Point>; // multidimensional
  public rotation: Property<number>;
  public opacity: Property<number>;

  public constructor() {
    this.anchorPoint = new Property<Point>(ZERO_POINT);
    this.position = new Property<Point>(ZERO_POINT);
    this.xPosition = new Property<number>(0);
    this.yPosition = new Property<number>(0);
    this.scale = new Property<Point>(new Point(1, 1));
    this.rotation = new Property<number>(0);
    this.opacity = new Property<number>(OPAQUE);
  }

  public excludeVaryingRanges(timeRanges: Array<TimeRange>): void {
    this.anchorPoint.excludeVaryingRanges(timeRanges);
    if (this.position !== undefined) {
      this.position.excludeVaryingRanges(timeRanges);
    } else {
      this.xPosition.excludeVaryingRanges(timeRanges);
      this.yPosition.excludeVaryingRanges(timeRanges);
    }
    this.scale.excludeVaryingRanges(timeRanges);
    this.rotation.excludeVaryingRanges(timeRanges);
    this.opacity.excludeVaryingRanges(timeRanges);
  }

  public gotoFrame(frame: number): void {
    this.anchorPoint.gotoFrame(frame);
    if (this.position !== undefined) {
      this.position.gotoFrame(frame);
    } else {
      this.xPosition.gotoFrame(frame);
      this.yPosition.gotoFrame(frame);
    }
    this.scale.gotoFrame(frame);
    this.rotation.gotoFrame(frame);
    this.opacity.gotoFrame(frame);
  }

  public verify(): boolean {
    return (
      this.anchorPoint !== undefined &&
      (this.position !== undefined || (this.xPosition !== undefined && this.yPosition !== undefined)) &&
      this.scale !== undefined &&
      this.rotation !== undefined &&
      this.opacity !== undefined
    );
  }
}

File: src/base/Transform2D.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////
//
//  Tencent is pleased to support the open source community by making libpag available.
//
//  Copyright (C) 2021 Tencent. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
//  except in compliance with the License. You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  unless required by applicable law or agreed to in writing, software distributed under the
//  license is distributed on an "as is" basis, without warranties or conditions of any kind,
//  either express or implied. see the license for the specific language governing permissions
//  and limitations under the license.
//
/////////////////////////////////////////////////////////////////////////////////////////////////

#include "base/utils/Verify.h"
#include "pag/file.h"

namespace pag {
std::unique_ptr<Transform2D> Transform2D::MakeDefault() {
  auto transform = new Transform2D();
  transform->anchorPoint = new Property<Point>();
  transform->anchorPoint->value = Point::Zero();
  transform->position = new Property<Point>();
  transform->position->value = Point::Zero();
  transform->scale = new Property<Point>();
  transform->scale->value = Point::Make(1, 1);
  transform->rotation = new Property<float>();
  transform->rotation->value = 0.0f;
  transform->opacity = new Property<Opacity>();
  transform->opacity->value = Opaque;
  return std::unique_ptr<Transform2D>(transform);
}
Transform2D::~Transform2D() {
  delete anchorPoint;
  delete position;
  delete xPosition;
  delete yPosition;
  delete scale;
  delete rotation;
  delete opacity;
}

void Transform2D::excludeVaryingRanges(std::vector<TimeRange>* timeRanges) const {
  anchorPoint->excludeVaryingRanges(timeRanges);
  if (position != nullptr) {
    position->excludeVaryingRanges(timeRanges);
  } else {
    xPosition->excludeVaryingRanges(timeRanges);
    yPosition->excludeVaryingRanges(timeRanges);
  }
  scale->excludeVaryingRanges(timeRanges);
  rotation->excludeVaryingRanges(timeRanges);
  opacity->excludeVaryingRanges(timeRanges);
}

bool Transform2D::verify() const {
  VerifyAndReturn(anchorPoint != nullptr &&
                  (position != nullptr || (xPosition != nullptr && yPosition != nullptr)) &&
                  scale != nullptr && rotation != nullptr && opacity != nullptr);
}
}  // namespace pag
